File: src/main/java/net/x3djsonld/data/Box.java
package net.x3djsonld.data;

import java.util.*;
import org.web3d.x3d.jsail.Core.*;
import org.web3d.x3d.jsail.fields.*;
import org.web3d.x3d.jsail.Geometry3D.*;
import org.web3d.x3d.jsail.Grouping.*;
import org.web3d.x3d.jsail.Navigation.*;
import org.web3d.x3d.jsail.Shape.*;

// Javadoc annotations follow, see below for source.
/**
 * <p> 3 boxes. </p>
 <p> Related links: Box.java source, <a href="http://www.web3d.org/x3d/content/examples/X3dResources.html" target="_blank">X3D Resources</a>, <a href="http://www.web3d.org/x3d/content/examples/X3dSceneAuthoringHints.html" target="_blank">X3D Scene Authoring Hints</a> and <a href="http://www.web3d.org/x3d/content/X3dTooltips.html" target="_blank">X3D Tooltips</a>. </p>
	<table style="color:black; border:0px solid; border-spacing:10px 0px;" summary="Scene Metadata">
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center; padding:10px 0px;"><i>meta tags</i></td>
			<td style="text-align:left;   padding:10px 0px;">net.x3djsonld.data.Box&nbsp; Document Metadata </td>
		</tr>

		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> title </i> </td>
			<td> <a href="https://coderextreme.net/X3DJSONLD/box.x3d">box.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> creator </i> </td>
			<td> John Carlson </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> generator </i> </td>
			<td> manual </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> identifier </i> </td>
			<td> <a href="https://coderextreme.net/X3DJSONLD/box.x3d" target="_blank">https://coderextreme.net/X3DJSONLD/box.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> description </i> </td>
			<td> 3 boxes </td>
		</tr>
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center;" colspan="2">  &nbsp; </td>
		</tr>
	</table>

	<p>
		This program uses the
		<a href="http://www.web3d.org/specifications/java/X3DJSAIL.html" target="_blank">X3D Java Scene Access Interface Library (X3DJSAIL)</a>.
		It has been produced using the 
		<a href="http://www.web3d.org/x3d/stylesheets/X3dToJava.xslt" target="_blank">X3dToJava.xslt</a>
		stylesheet to create Java source code from an <code>.x3d</code> scene.
	</p>

	* @author John Carlson
 */

public class Box
{
  /** Default constructor to create this object. */
  public Box ()
  {
    initialize();
  }
	
  /** Create and initialize the X3D model. */
  public final void initialize()
  {
  x3dModel = new X3DObject().setProfile("Immersive").setVersion("3.3")
  .setHead(new headObject()
    .addMeta(new metaObject().setName("title").setContent("box.x3d"))
    .addMeta(new metaObject().setName("creator").setContent("John Carlson"))
    .addMeta(new metaObject().setName("generator").setContent("manual"))
    .addMeta(new metaObject().setName("identifier").setContent("https://coderextreme.net/X3DJSONLD/box.x3d"))
    .addMeta(new metaObject().setName("description").setContent("3 boxes")))
  .setScene(new SceneObject()
    .addChild(new NavigationInfoObject().setType(new MFStringObject("\"EXAMINE\"")))
    .addChild(new ViewpointObject().setDescription("Cubes on Fire").setPosition(0.0f,0.0f,12.0f))
    .addChild(new ProtoDeclareObject().setName("anyShape")
      .setProtoInterface(new ProtoInterfaceObject()
        .addField(new fieldObject().setAccessType("inputOutput").setName("xtranslation").setType("SFVec3f").setValue("0 0 0"))
        .addField(new fieldObject().setAccessType("inputOutput").setName("myShape").setType("MFNode")
          .addChild(new ShapeObject()
            .setGeometry(new SphereObject())
            .setAppearance(new AppearanceObject()
              .setMaterial(new MaterialObject().setDiffuseColor(1.0f,1.0f,1.0f))))))
      .setProtoBody(new ProtoBodyObject()
        .addChild(new TransformObject()
          .setIS(new ISObject()
            .addConnect(new connectObject().setNodeField("translation").setProtoField("xtranslation"))
            .addConnect(new connectObject().setNodeField("children").setProtoField("myShape"))))))
    .addChild(new ProtoDeclareObject().setName("three")
      .setProtoInterface(new ProtoInterfaceObject()
        .addField(new fieldObject().setAccessType("inputOutput").setName("ytranslation").setType("SFVec3f").setValue("0 0 0"))
        .addField(new fieldObject().setAccessType("inputOutput").setName("myShape").setType("MFNode")
          .addChild(new ShapeObject()
            .setGeometry(new CylinderObject())
            .setAppearance(new AppearanceObject()
              .setMaterial(new MaterialObject().setDiffuseColor(1.0f,1.0f,1.0f))))))
      .setProtoBody(new ProtoBodyObject()
        .addChild(new TransformObject()
          .setIS(new ISObject()
            .addConnect(new connectObject().setNodeField("translation").setProtoField("ytranslation")))
          .addChild(new ProtoInstanceObject().setName("anyShape")
            .addFieldValue(new fieldValueObject().setName("xtranslation").setValue("0 0 0"))
            .setIS(new ISObject()
              .addConnect(new connectObject().setNodeField("myShape").setProtoField("myShape"))))
          .addChild(new ProtoInstanceObject().setName("anyShape")
            .addFieldValue(new fieldValueObject().setName("xtranslation").setValue("2 0 0"))
            .setIS(new ISObject()
              .addConnect(new connectObject().setNodeField("myShape").setProtoField("myShape"))))
          .addChild(new ProtoInstanceObject().setName("anyShape")
            .addFieldValue(new fieldValueObject().setName("xtranslation").setValue("-2 0 0"))
            .setIS(new ISObject()
              .addConnect(new connectObject().setNodeField("myShape").setProtoField("myShape")))))))
    .addChild(new ProtoInstanceObject().setName("three")
      .addFieldValue(new fieldValueObject().setName("ytranslation").setValue("0 0 0"))
      .addFieldValue(new fieldValueObject().setName("myShape")
        .addChild(new ShapeObject()
          .setGeometry(new BoxObject().setSize(1.0f,1.0f,1.0f))
          .setAppearance(new AppearanceObject()
            .setMaterial(new MaterialObject().setDiffuseColor(0.0f,1.0f,0.0f)))))));
  }
  // end of initialize() method

  /** The initialized model object, created within initialize() method. */
  private X3DObject x3dModel;
  
  /** Provide a shallow copy of the X3D model.
   * @return Box model
   */
  public X3DObject getX3dModel()
  {	  
	  return x3dModel;
  }
  
  /** Indicate X3DJSAIL validation results for this X3D model.
   * @return validation results plus exception information, if any
   */
  public String validateSelf()
  {
	String       metaResult = new String();
	String validationResult = new String();
	String  exceptionResult = new String();
	try
	{
		initialize();
		
		if ((getX3dModel() == null) || (getX3dModel().getHead() == null))
		{
			validationResult = "empty scene, nothing to validate. " + x3dModel.validate();
			return validationResult;
		}
		// first list informational meta elements of interest
		for (metaObject meta : getX3dModel().getHead().getMetaList())
		{
			if (meta.getName().equals(metaObject.NAME_ERROR) ||
				meta.getName().equals(metaObject.NAME_WARNING) ||
				meta.getName().equals(metaObject.NAME_HINT) ||
				meta.getName().equals(metaObject.NAME_INFO) ||
				meta.getName().equals(metaObject.NAME_TODO))
			{
				metaResult += meta.toStringX3D();
			}
		}
		validationResult += x3dModel.validate(); // walk entire tree to validate correctness
	}
	catch (Exception e)
	{
		exceptionResult = e.getMessage(); // report exception failures, if any
	    if (exceptionResult == null)
	    {
			exceptionResult = "Exception caught but null message!";
			e.printStackTrace();
	    }
	}
	if  (metaResult.isEmpty() && exceptionResult.isEmpty() && validationResult.isEmpty())
	     return "success";
	else
	{
		String returnMessage = metaResult;
		if  (!exceptionResult.isEmpty() && !validationResult.isEmpty())
			returnMessage += "\n*** ";
		returnMessage += exceptionResult;
		if  (exceptionResult.isEmpty() && !validationResult.isEmpty())
			returnMessage = "\n" + returnMessage; // skip line before meta tags, etc.
		returnMessage += validationResult;
		return returnMessage;
	}
  }
    /** Default main() method provided for test purposes.
     * @param argv input parameters
     */
    public static void main(String argv[])
    {
		Box testObject = new Box();
		System.out.print("Box execution self-validation test results: ");
		String validationResults = testObject.validateSelf();
		if (validationResults.startsWith("<"))
			System.out.println();
		System.out.println(validationResults);
	}
}

File: Box.py
from jnius import autoclass
from X3Dautoclass import *
X3D0 =  X3DObject().setProfile("Immersive").setVersion("3.3")
head1 = headObject()
meta2 = metaObject().setName("title").setContent("box.x3d")
head1.addMeta(meta2)
meta3 = metaObject().setName("creator").setContent("John Carlson")
head1.addMeta(meta3)
meta4 = metaObject().setName("generator").setContent("manual")
head1.addMeta(meta4)
meta5 = metaObject().setName("identifier").setContent("http://coderextreme.net/X3DJSONLD/box.x3d")
head1.addMeta(meta5)
meta6 = metaObject().setName("description").setContent("3 boxes")
head1.addMeta(meta6)
meta7 = metaObject().setName("translated").setContent("15 May 2017")
head1.addMeta(meta7)
meta8 = metaObject().setName("generator").setContent("X3dToJson.xslt, http://www.web3d.org/x3d/stylesheets/X3dToJson.html")
head1.addMeta(meta8)
meta9 = metaObject().setName("reference").setContent("X3D JSON encoding: http://www.web3d.org/wiki/index.php/X3D_JSON_Encoding")
head1.addMeta(meta9)
meta10 = metaObject().setName("translated").setContent("15 May 2017")
head1.addMeta(meta10)
meta11 = metaObject().setName("generator").setContent("X3DJSONLD: https://github.com/coderextreme/X3DJSONLD")
head1.addMeta(meta11)
X3D0.setHead(head1)
Scene12 = SceneObject()
NavigationInfo13 = NavigationInfoObject().setType(["EXAMINE"])
Scene12.addChild(NavigationInfo13)
Viewpoint14 = ViewpointObject().setDescription("Cubes on Fire").setPosition([0,0,12])
Scene12.addChild(Viewpoint14)
ProtoDeclare15 = ProtoDeclareObject().setName("anyShape")
ProtoInterface16 = ProtoInterfaceObject()
field17 = fieldObject().setType(fieldObject.TYPE_SFVEC3F).setName("xtranslation").setAccessType(fieldObject.ACCESSTYPE_INPUTOUTPUT).setValue("0 0 0")
ProtoInterface16.addField(field17)
field18 = fieldObject().setType(fieldObject.TYPE_MFNODE).setName("myShape").setAccessType(fieldObject.ACCESSTYPE_INPUTOUTPUT)
Shape19 = ShapeObject()
Sphere20 = SphereObject()
Shape19.setGeometry(Sphere20)
Appearance21 = AppearanceObject()
Material22 = MaterialObject().setDiffuseColor([1,1,1])
Appearance21.setMaterial(Material22)
Shape19.setAppearance(Appearance21)
field18.addChild(Shape19)
ProtoInterface16.addField(field18)
ProtoDeclare15.setProtoInterface(ProtoInterface16)
ProtoBody23 = ProtoBodyObject()
Transform24 = TransformObject()
IS25 = ISObject()
connect26 = connectObject().setNodeField("translation").setProtoField("xtranslation")
IS25.addConnect(connect26)
connect27 = connectObject().setNodeField("children").setProtoField("myShape")
IS25.addConnect(connect27)
Transform24.setIS(IS25)
ProtoBody23.addChild(Transform24)
ProtoDeclare15.setProtoBody(ProtoBody23)
Scene12.addChild(ProtoDeclare15)
ProtoDeclare28 = ProtoDeclareObject().setName("three")
ProtoInterface29 = ProtoInterfaceObject()
field30 = fieldObject().setType(fieldObject.TYPE_SFVEC3F).setName("ytranslation").setAccessType(fieldObject.ACCESSTYPE_INPUTOUTPUT).setValue("0 0 0")
ProtoInterface29.addField(field30)
field31 = fieldObject().setType(fieldObject.TYPE_MFNODE).setName("myShape").setAccessType(fieldObject.ACCESSTYPE_INPUTOUTPUT)
Shape32 = ShapeObject()
Cylinder33 = CylinderObject()
Shape32.setGeometry(Cylinder33)
Appearance34 = AppearanceObject()
Material35 = MaterialObject().setDiffuseColor([1,1,1])
Appearance34.setMaterial(Material35)
Shape32.setAppearance(Appearance34)
field31.addChild(Shape32)
ProtoInterface29.addField(field31)
ProtoDeclare28.setProtoInterface(ProtoInterface29)
ProtoBody36 = ProtoBodyObject()
Transform37 = TransformObject()
IS38 = ISObject()
connect39 = connectObject().setNodeField("translation").setProtoField("ytranslation")
IS38.addConnect(connect39)
Transform37.setIS(IS38)
ProtoInstance40 = ProtoInstanceObject().setName("anyShape")
fieldValue41 = fieldValueObject().setName("xtranslation").setValue("0 0 0")
ProtoInstance40.addFieldValue(fieldValue41)
IS42 = ISObject()
connect43 = connectObject().setNodeField("myShape").setProtoField("myShape")
IS42.addConnect(connect43)
ProtoInstance40.setIS(IS42)
Transform37.addChild(ProtoInstance40)
ProtoInstance44 = ProtoInstanceObject().setName("anyShape")
fieldValue45 = fieldValueObject().setName("xtranslation").setValue("2 0 0")
ProtoInstance44.addFieldValue(fieldValue45)
IS46 = ISObject()
connect47 = connectObject().setNodeField("myShape").setProtoField("myShape")
IS46.addConnect(connect47)
ProtoInstance44.setIS(IS46)
Transform37.addChild(ProtoInstance44)
ProtoInstance48 = ProtoInstanceObject().setName("anyShape")
fieldValue49 = fieldValueObject().setName("xtranslation").setValue("-2 0 0")
ProtoInstance48.addFieldValue(fieldValue49)
IS50 = ISObject()
connect51 = connectObject().setNodeField("myShape").setProtoField("myShape")
IS50.addConnect(connect51)
ProtoInstance48.setIS(IS50)
Transform37.addChild(ProtoInstance48)
ProtoBody36.addChild(Transform37)
ProtoDeclare28.setProtoBody(ProtoBody36)
Scene12.addChild(ProtoDeclare28)
ProtoInstance52 = ProtoInstanceObject().setName("three")
fieldValue53 = fieldValueObject().setName("ytranslation").setValue("0 0 0")
ProtoInstance52.addFieldValue(fieldValue53)
fieldValue54 = fieldValueObject().setName("myShape")
Shape55 = ShapeObject()
Box56 = BoxObject().setSize([1,1,1])
Shape55.setGeometry(Box56)
Appearance57 = AppearanceObject()
Material58 = MaterialObject().setDiffuseColor([0,1,0])
Appearance57.setMaterial(Material58)
Shape55.setAppearance(Appearance57)
fieldValue54.addChild(Shape55)
ProtoInstance52.addFieldValue(fieldValue54)
Scene12.addChild(ProtoInstance52)
X3D0.setScene(Scene12)

X3D0.toFileX3D("Box.new.x3d")
